pkg/util/region_range_lock.go

```diff
@@ -0,0 +1,337 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"math"
+	"sync"
+	"sync/atomic"
+
+	"github.com/google/btree"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/store/tikv/oracle"
+	"go.uber.org/zap"
+)
+
+type rangeTsEntry struct {
+	// Only startKey is necessary. End key can be inferred by the next item since the map always keeps a continuous
```

Comment:
What if there is a gap between two adjacent items?

Reply:
When updating a range, if there is not another entry with it's key = `endKey`, it will be inserted to keep the value of the original value of the range after the new range. For example, 
* "a" -> 1 (represents ["a", "d") is 1)
* "d" -> 2
if we set ["b", "c") to 10, it will become:
* "a" -> 1 (represents ["a", "b") is 1)
* "b" -> 10 (represents ["b", "c") is 10)
* "c" -> 1 (represents ["c", "d") is 1)
* "d" -> 2 
And if there's no value set in some range, the default value will be considered MaxUint64 (because we need to calculate the minimal value among ranges)